### ISSUE TITLE
Read camtrap-dp data with camtrapR

### ISSUE THREAD

**Issue by damianooldoni** [jniedballa/camtrapR#14]
[Camera Trap Data Package (Camtrap DP)](https://gitlab.com/oscf/camtrap-dp) has been greatly developed in the last months. And it will become a [TDWG](https://www.tdwg.org/) standard (see https://gitlab.com/oscf/camtrap-dp/-/issues/85) soon as it is mature enough. This means the use of this data format will greatly increase in the near future. [Agouti](https://www.agouti.eu/) uses already this format to export its data.

What do you think about a read function in `camtrapR` for camtrap-dp formatted data? Is it already part of your dev branch? If not, I and other INBO colleagues (@jimcasaer and its colleagues of the Fauna and Invasive Species team) we are quit interested to work on this. Thanks.

**Comment by damianooldoni** [comment]
Update: camtrap-dp is already a TDWG standard. See https://tdwg.github.io/camtrap-dp/

**Comment by jniedballa** [comment]
Hi, 
is the definition of the standard finished? And do you expect the standard to evolve further in the future, requiring ongoing updates?
So far I didn't do any work in this direction within camtrapR, but would of course be open to discuss how to approach this. From what I see there are some key differences:

- camtrapR combines the content of multimedia.csv and observations.csv in one record table
- camtrapR doesn't require a metadata table
- the format of the deployment table differs, in particular the use of the Problem columns in the camera trap tables of camtrapR is not compatible with the standard, but essential to many function within camtrapR. 

Right now I am hesitant to change core functionalities of camtrapR that many people got used to and have code for. So I'd suggest we find a way to convert the data automatically between the two.

There might be more issues when digging into the details. But I'm sure we can write an export function that takes care of these things. 
Anyways, I'm happy to discuss about this.

Cheers,
Jürgen

**Comment by jimcasaer** [comment]
Hi Jürgen,

I fully agree that you should not change the core functionalities of camtrapR given the number of users of the package at the moment. I think the way forward is that we look together how we can import/translate the data from deployments / observations of the new export format to the existing objects in camtrapR and add new functionalities strating from existing objects in camtrapR where possible. 

I will share a google-doc with you next week that is a working document so you can see the total set of data-exploration and data-visualisation tools we are aiming for.

have a nice weekend,
Jim

**Comment by damianooldoni** [comment]
Indeed, @jniedballa: the point is to allow users to read data from camtrap-dp standard in camtrapR, so they can use camtrapR package to analyze them. The way data are stored in camtrapR should not change unless you see the need for it in the future.

About standard stability, I think it is quite stable up to now, However, it is far from being a mature standard. @peterdesmet: as you are actively involved in the standard development/maintenance, can you confirm this?

Thanks.

**Comment by peterdesmet** [comment]
Indeed, even though camtrap-dp is now hosted under Biodiversity Information Standards (TDWG), it is not a standard yet. Pending one issue with the metadata, it is a 1.0 release. @kbubnicki and I have worked hard to get it there and it is now the only export format for https://agouti.eu, meaning many users will get their hands on it. That is a good thing, because it brings the format from theory into practice, and test how well it meets different use cases. It also means fields might get added in the future.

What would be really helpful to users though is that some of the complexity/development is hidden to them and they can just read the format with camtrapR and work with the camtrapR objects they are familiar with (as mentioned by @damianooldoni and @jimcasaer above). That would also avoid users/developers writing their own functions to work with the format.

**Comment by kbubnicki** [comment]
Hi all!
I also think that the implementation of some user-friendly import/export interface (2 functions?) for reading/writing camtrap-dp data packages directly to/from camtrapR would be wonderful, especially thinking about how popular and useful is camtrapR already! Many of our [TRAPPER](https://gitlab.com/oscf/trapper-project) users are very interested in this integration. It could also help with building semi-automatic camera trapping data analysis pipelines based on e.g. Agouti and TRAPPER as data providers (via API) and camtrapR as a direct interface to the R analytical environment.

**Comment by jniedballa** [comment]
Hi, 
sure, sound good to me. Generally an import / export functionality sounds like a reasonable thing to implement.
I'll wait for the GDoc Jim mentioned, then we can discuss in more detail what is required.

**Comment by guswiler** [comment]
Hi all, as it's been a few years, I was curious if there has been any further development on the readability of camtrap-dp data within camptrapR?

**Comment by kbubnicki** [comment]
Good question, It would be nice to know that @jniedballa ? :)

**Comment by jniedballa** [comment]
Hi all, 
Please have a look at the current dev version here on GitHub, which has a [readcamtrapDP() ](https://github.com/jniedballa/camtrapR/blob/dev/R/readcamtrapDP.R)function. It handles the data import and modifies the format slightly to match the camtrapR standard. It is fully integrated with the shiny dashboard too, so people can import, visualize and analyze data all in one place (also works for Wildlife Insights data and generic csv files). The dashboard is accessible through surveyDashboard(). But of course the data can be used in the normal code-based camtrapR workflow as well.

There is no capability to write camtrap DP data yet. 

Install the dev version via:
```
# install.packages("remotes") 
# install.packages("R.rsp")
remotes::install_github("jniedballa/camtrapR", ref = "dev", build_vignettes = TRUE)
```

I am aware of the camtrapdp package on CRAN and wouldn't mind using the read and validate functionality provided there instead in the future. As we discussed before, a single central import function for camtrap DP data in R makes most sense. 
 
Feel free to give it a try and provide feedback.

**Comment by MartijnUH** [comment]
> Feel free to give it a try and provide feedback.

Hi @jniedballa,

Great to see there's progress towards integrating `camtrap DP` and `camtrapR` . 
I am happy to try things out and provide you with feedback. However, when trying to install the development version of `camtrapR`, I receive a warning message that the exports `PPC.community, PPC.residuals` are undefined:

```{r}
Installing package into ‘C:/R/library’
(as ‘lib’ is unspecified)
* installing *source* package 'camtrapR' ...
** using staged installation
** R
** data
** inst
** byte-compile and prepare package for lazy loading
** help
*** installing help indices
** building package indices
** installing vignettes
** testing if installed package can be loaded from temporary location
Error: package or namespace load failed for 'camtrapR' in namespaceExport(ns, exports):
 undefined exports: PPC.community, PPC.residuals
Error: loading failed
Execution halted
ERROR: loading failed
* removing 'C:/R/library/camtrapR'
Warning in install.packages :
  installation of package ‘C:/Users/martijn_bollen/Downloads/camtrapR-dev’ had non-zero exit status
```

I tried installing the package from GitHub, and from source, after downloading the dev-repository:

```{r}
remotes::install_github("jniedballa/camtrapR", ref = "dev", build_vignettes = TRUE)
remotes::install_github("jniedballa/camtrapR", ref = "dev", build_vignettes = FALSE)
remotes::install_github("jniedballa/camtrapR@dev", force = TRUE)
install.packages("C:/Users/martijn_bollen/Downloads/camtrapR-dev", repos = NULL, type = "source")
```

Could you check if `PPC.community, PPC.residuals` exist in the dev-branch? Or am I missing something?

Thanks for your help!

Martijn

**Comment by jniedballa** [comment]
apologies, I forgot that. Removed it for now. Please try again.

**Comment by MartijnUH** [comment]
> apologies, I forgot that. Removed it for now. Please try again. Those are functions for posterior predictive checks in community occupancy models that need a little more validation and will be added next week.

It works now. Thanks for your swift reply.

**Comment by peterdesmet** [comment]
That's good news! Regarding:

> I am aware of the camtrapdp package on CRAN and wouldn't mind using the read and validate functionality provided there instead in the future. As we discussed before, a single central import function for camtrap DP data in R makes most sense.

I would recommend doing so.

- By relying on `camtrapdp::read_camtrapdp()`, older datasets will be upgraded automatically to the latest version of the Camtrap DP standard, making camtrapR compatible with current and future datasets. This is not an issue now (there is only one version of the standard) but it will become useful in the future.
- If you store a dataset as a camtrapdp object, you can use camtrapdp's [print, check](https://inbo.github.io/camtrapdp/reference/index.html#miscellaneous-functions), [filter](https://inbo.github.io/camtrapdp/reference/index.html#filter-functions) and [accessor and assignment](https://inbo.github.io/camtrapdp/reference/index.html#accessor-and-assignment-functions) functions. It also has a [write_camtrapdp()](https://inbo.github.io/camtrapdp/reference/write_camtrapdp.html).
- Centralizing Camtrap DP handling (read, print, validate, write) in camtrapdp = less duplication of effort. Note that validation is currently pretty minimal, but there are [plans](https://github.com/inbo/camtrapdp/issues/58) to extend it.
